packages/contracts-bedrock/src/l1/SuperchainConfig.sol

```diff
@@ -0,0 +1,247 @@
+    ///         It can only be modified by an upgrade.
+    bytes32 public constant VETOER_SLOT = bytes32(uint256(keccak256("superchainConfig.vetoer")) - 1);
+
+    /// @notice The address of the guardian, can pause the OptimismPortal.
+    ///         It can only be modified by an upgrade.
+    bytes32 public constant GUARDIAN_SLOT = bytes32(uint256(keccak256("superchainConfig.guardian")) - 1);
+
+    /// @notice The delay time in seconds between when an upgrade is initiated and when it can be finalized.
+    ///         It can only be modified by an upgrade.
+    bytes32 public constant DELAY_SLOT = bytes32(uint256(keccak256("superchainConfig.delay")) - 1);
+
+    /// @notice The pause status of withdrawals from an chain in the superchain.
+    ///         Set by the guardian role.
+    bytes32 public constant PAUSED_SLOT = bytes32(uint256(keccak256("superchainConfig.paused")) - 1);
+
+    /// @notice Mapping of allowed sequencers.
+    ///         The initiator should be able to add to it instantly, but removing is subject to delay.
+    mapping(bytes32 => bool) public allowedSequencers;
+
+    /// @notice Emitted when the pause is triggered.
+    event Paused();
+
+    /// @notice Emitted when the pause is lifted.
+    event Unpaused();
+
+    /// @notice Emitted when configuration is updated.
+    /// @param updateType Type of update.
+    /// @param data       Encoded update data.
+    event ConfigUpdate(UpdateType indexed updateType, bytes data);
+
+    /// @notice Semantic version.
+    /// @custom:semver 1.0.0
+    string public constant version = "1.0.0";
+
+    /// @notice Constructs the SuperchainConfig contract.
+    constructor() {
+        initialize({
+            _systemOwner: address(0),
+            _initiator: address(0),
+            _vetoer: address(0),
+            _guardian: address(0),
+            _delay: 0,
+            _sequencers: new Types.SequencerKeys[](0)
+        });
+    }
+
+    /// @notice Initializer.
+    ///         The resource config must be set before the require check.
+    /// @param _systemOwner     Initial owner of the contract.
+    /// @param _initiator Address of the initiator who may initiate an upgrade or change to critical config values.
+    /// @param _vetoer    Address of the vetoer.
+    /// @param _guardian  Address of the guardian, can pause the OptimismPortal.
+    /// @param _delay     The delay time in seconds between when an upgrade is initiated and when it can be finalized.
+    /// @param _sequencers The initial list of allowed sequencers
+    function initialize(
+        address _systemOwner,
+        address _initiator,
+        address _vetoer,
+        address _guardian,
+        uint256 _delay,
+        Types.SequencerKeys[] memory _sequencers
+    )
+        public
+        reinitializer(2)
+    {
+        _setAddress(SYSTEM_OWNER_SLOT, _systemOwner);
+        emit ConfigUpdate(UpdateType.SYSTEM_OWNER, abi.encode(_systemOwner));
```

Comment:
Prefer internal functions and not emitting the events directly,

```solidity
function _setSystemOwner(address _systemOwner) internal {
  _setAddress(SYSTEM_OWNER_SLOT, _systemOwner);
  emit ConfigUpdate(...);
}
```

Reply:
Can do, note however that these variables (with the exception of `paused`) are all updated via upgrade, so we'd be creating 6 new setters that are only used in the initializer.

Reply:
Did this.